### PR TITLE
[Stacked 3] Add storage disk to Size Variants.

### PR DESCRIPTION
**Pull request by ildyria**

### CODE REVIEW

File: app/Models/SizeVariant.php [hunk 1]
```diff
@@ -118,6 +108,7 @@ class SizeVariant extends Model
 		'photo_id', // see above
 		'short_path',  // serialize url instead
 		'sym_links', // don't serialize relation of symlinks
+		'',
```
Comment (ildyria) [inline app/Models/SizeVariant.php]: todo: nuke this.

File: app/Models/SizeVariant.php [hunk 2]
```diff
@@ -183,39 +181,68 @@ public function sym_links(): HasManyBidirectionally
...
 	public function getUrlAttribute(): string
 	{
-		$imageDisk = AbstractSizeVariantNamingStrategy::getImageDisk();
+		$imageDisk = Storage::disk($this->storage_disk->value);
 
 		if (
-			(Auth::user()?->may_administrate === true && !Configs::getValueAsBool('SL_for_admin')) ||
-			!Configs::getValueAsBool('SL_enable')
+			!Configs::getValueAsBool('SL_enable') ||
+			(!Configs::getValueAsBool('SL_for_admin') && Auth::user()?->may_administrate === true)
```
Comment (d7415) [inline app/Models/SizeVariant.php:187]: :+1:

File: app/Models/SizeVariant.php [hunk 3]
```diff
@@ -183,39 +181,68 @@ public function sym_links(): HasManyBidirectionally
...
 		// In order to allow a grace period, we create a new symbolic link,
-		// if the most recent existing link has reached 2/3 of its lifetime
 		$maxLifetime = Configs::getValueAsInt('SL_life_time_days') * 24 * 60 * 60;
-		$gracePeriod = $maxLifetime / 3;
+		$imageDisk = Storage::disk($this->storage_disk->value);
 
-		$storageAdapter = $imageDisk->getAdapter();
+		// Return the public URL in case the S3 bucket is set to public, otherwise generate a temporary URL
+		$visibility = config('filesystems.disks.s3.visibility', 'private');
+		// $visibility = $imageDisk->getConfig()['visibility'] ?? 'private';
```
Comment (d7415) [inline app/Models/SizeVariant.php]:
```suggestion
```
?

File: app/Models/SizeVariant.php [hunk 4]
```diff
@@ -183,39 +181,68 @@ public function sym_links(): HasManyBidirectionally
...
+		if ($visibility === 'public') {
+			return $imageDisk->url($this->short_path);
```
Comment (d7415) [inline app/Models/SizeVariant.php:218]: I guess at some point we'll need to document that the bucket has to be private to used the symlink options.

**Review by d7415 — APPROVED** [review]

File: app/Models/SizeVariant.php (hunk 1 above)
Comment (ildyria) [reply app/Models/SizeVariant.php]: [fc7a773](https://github.com/LycheeOrg/Lychee/pull/2381/commits/fc7a773d63ca0d37bb8a77f8f4c08bd49ef31023)

File: app/Models/SizeVariant.php (hunk 3 above)
Comment (ildyria) [reply app/Models/SizeVariant.php]: [9b1a7be](https://github.com/LycheeOrg/Lychee/pull/2381/commits/9b1a7be561d015641c19c3d698bdd303f075cb60)

File: app/Models/SizeVariant.php (hunk 4 above)
Comment (ildyria) [reply app/Models/SizeVariant.php:218]: Indeed.

**Pull request merged**